File: week1_baseline/python/09_global_executable/lib/boukensha/run_dsl.py
class RunDSL:
    # Passed to a `boukensha.run` `register` callback. Exposes only `tool`,
    # keeping the DSL surface intentionally small.
    def __init__(self, registry):
        self._registry = registry

    def tool(self, name, *, description, parameters=None):
        return self._registry.tool(name, description=description, parameters=parameters)

File: week1_baseline/rust/07_the_run_dsl/src/run_dsl.rs
use std::collections::HashMap;

use serde_yaml_ng::Value;

use crate::registry::Registry;
use crate::tasks::Task;

/// Passed to `run`'s `register` callback. Exposes only `tool`, keeping the
/// DSL surface intentionally small.
pub struct RunDsl<'a, T: Task> {
    registry: &'a mut Registry<T>,
}

impl<'a, T: Task> RunDsl<'a, T> {
    pub fn new(registry: &'a mut Registry<T>) -> Self {
        Self { registry }
    }

    pub fn tool(
        &mut self,
        name: impl Into<String>,
        description: impl Into<String>,
        parameters: Value,
        block: impl Fn(&HashMap<String, String>) -> Result<String, String> + 'static,
    ) {
        self.registry.tool(name, description, parameters, block);
    }
}
